internal/pally/value.go

```diff
@@ -0,0 +1,105 @@
+package pally
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+const _defaultVectorSize = 100
+
+// Value is an atomic with some associated metadata. It's a building block
+// for higher-level metric types.
+type value struct {
+	atomic.Int64
+
+	opts              Opts
+	desc              *prometheus.Desc
+	variableLabelVals []string
+}
+
+func newValue(opts Opts) value {
+	return value{opts: opts, desc: opts.describe()}
+}
+
+// Describe implements half of prometheus.Collector.
+func (v value) Describe(ch chan<- *prometheus.Desc) {
+	ch <- v.desc
+}
+
+// A vector is a collection of values that share the same metadata.
+type vector struct {
+	opts Opts
+	desc *prometheus.Desc
+
+	// The factory function creates a new metric given the vector's metadata
+	// and values for the variable labels.
+	factory func(Opts, *prometheus.Desc, []string) metric
+
+	metricsMu sync.RWMutex
+	metrics   map[string]metric
+}
+
+func (vec *vector) getOrCreate(labels ...string) (metric, error) {
+	for _, l := range labels {
+		if !isValidTallyString(l) {
+			return nil, errors.New("variable label values must also be Tally-compatible")
+		}
+	}
+
+	digester := newDigester()
+	for _, s := range labels {
+		digester.add(s)
+	}
+
+	vec.metricsMu.RLock()
+	m, ok := vec.metrics[string(digester.digest())]
+	vec.metricsMu.RUnlock()
+	if ok {
+		digester.free()
+		return m, nil
+	}
+
+	vec.metricsMu.Lock()
+	m, err := vec.newValue(digester.digest(), labels)
```

Comment:
need to recheck `vec.metrics[string(digester.digest())]` under the lock. it's possible a new counter was created between the `RUnlock` and the `Lock`